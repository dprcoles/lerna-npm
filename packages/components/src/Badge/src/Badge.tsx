import cn from "classnames"
import { cva, type VariantProps } from "cva"
import * as React from "react"

export enum BadgeVariant {
  Primary = "primary",
  Secondary = "secondary",
  Danger = "danger",
  Outline = "outline",
}

const badgeVariants = cva("dc-c-badge", {
  variants: {
    variant: {
      [BadgeVariant.Primary]: "dc-c-badge--primary",
      [BadgeVariant.Secondary]: "dc-c-badge--secondary",
      [BadgeVariant.Danger]: "dc-c-badge--danger",
      [BadgeVariant.Outline]: "dc-c-badge--outline",
    },
  },
  defaultVariants: {
    variant: BadgeVariant.Primary,
  },
})

export type BadgeProps<E extends React.ElementType> = {
  as?: E
  children?: React.ReactNode
} & React.ComponentPropsWithoutRef<E> &
  VariantProps<typeof badgeVariants>

const Badge = ({ as = "div", className, variant, ...props }: BadgeProps<React.ElementType>) => {
  const Component = as
  return <Component className={cn(badgeVariants({ variant }), className)} {...props} />
}

export { Badge, badgeVariants }
