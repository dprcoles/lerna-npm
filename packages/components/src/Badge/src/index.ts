export type { BadgeProps } from "./Badge"
export { Badge, BadgeVariant } from "./Badge"
