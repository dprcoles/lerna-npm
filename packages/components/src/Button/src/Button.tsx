import cn from "classnames"
import { cva, type VariantProps } from "cva"
import * as React from "react"

export enum ButtonVariant {
  Primary = "primary",
  Secondary = "secondary",
  Danger = "danger",
  Outline = "outline",
  Ghost = "ghost",
  Link = "link",
}

export enum ButtonSize {
  Default = "default",
  Small = "small",
  Large = "large",
  Icon = "icon",
}

const buttonVariants = cva("dc-c-button", {
  variants: {
    variant: {
      [ButtonVariant.Primary]: "dc-c-button--primary",
      [ButtonVariant.Secondary]: "dc-c-button--secondary",
      [ButtonVariant.Danger]: "dc-c-button--danger",
      [ButtonVariant.Outline]: "dc-c-button--outline",
      [ButtonVariant.Ghost]: "dc-c-button--ghost",
      [ButtonVariant.Link]: "dc-c-button--link",
    },
    size: {
      [ButtonSize.Default]: "",
      [ButtonSize.Small]: "dc-c-button--small",
      [ButtonSize.Large]: "dc-c-button--large",
      [ButtonSize.Icon]: "dc-c-button--icon",
    },
  },
  defaultVariants: {
    variant: ButtonVariant.Primary,
    size: ButtonSize.Default,
  },
})

export type ButtonProps<E extends React.ElementType> = {
  as?: E
  children?: React.ReactNode
  isFullWidth?: boolean
} & React.ComponentPropsWithoutRef<E> &
  VariantProps<typeof buttonVariants>

const Button = React.forwardRef<
  HTMLButtonElement | HTMLAnchorElement,
  ButtonProps<React.ElementType>
>(({ as = "button", className, variant, isFullWidth, size, ...props }, ref) => {
  const Component = as
  return (
    <Component
      className={cn(
        buttonVariants({ variant, size, className }),
        isFullWidth && "dc-c-button--full"
      )}
      ref={ref}
      {...props}
    />
  )
})
Button.displayName = "Button"

export { Button, buttonVariants }
