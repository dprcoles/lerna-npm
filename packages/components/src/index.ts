export * from "./Badge/src"
export * from "./Button/src"
