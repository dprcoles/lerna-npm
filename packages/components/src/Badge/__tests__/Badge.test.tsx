import { render } from "@testing-library/react"
import React from "react"

import { Badge, BadgeVariant } from "../src/Badge"

const BADGE_TEXT = "Test Badge"

describe("<Badge />", () => {
  const setup = (propOverrides = {}) => {
    const props = {
      onClick: jest.fn(),
      ...propOverrides,
    }

    const wrapper = render(<Badge {...props}>{BADGE_TEXT}</Badge>)

    return {
      ...wrapper,
      base: wrapper.container.querySelector(".dc-c-badge")!,
    }
  }

  it("shows the expected badge text", () => {
    const { base } = setup()

    expect(base).toHaveTextContent(BADGE_TEXT)
  })

  it("adds any expected class names", () => {
    const { base } = setup({ className: "test-class" })

    expect(base).toHaveClass("test-class")
  })

  it("allows the use of an alternative element", () => {
    const { base } = setup({ as: "a" })

    expect(base).toHaveClass("dc-c-badge")
    expect(base.tagName).toBe("A")
  })

  it.each([...Object.values(BadgeVariant)])(
    "adds the expected %s variant class name modifier",
    variant => {
      const { base } = setup({ variant })

      expect(base).toHaveClass(`dc-c-badge--${variant}`)
    }
  )

  it("renders matching snapshot", () => {
    const { asFragment } = setup()

    expect(asFragment()).toMatchSnapshot()
  })
})
