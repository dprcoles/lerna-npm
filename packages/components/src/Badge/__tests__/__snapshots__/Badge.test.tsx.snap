// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Badge /> renders matching snapshot 1`] = `
<DocumentFragment>
  <div
    class="dc-c-badge dc-c-badge--primary"
  >
    Test Badge
  </div>
</DocumentFragment>
`;
